### ISSUE TITLE
[css-text-3] The spec should provide guidance about interop issues related to ink overflow from text

### ISSUE THREAD

**Issue by szager-chromium** [w3c/csswg-drafts#10066]
This is an offshoot from #8649. In cases where the extent of ink overflow can't be pre-determined or programmatically measured (primarily text and text decorations), we should offer non-conforming guidance to authors about interoperability issues and provide best practices for minimizing them.

This would be generally useful information, and would be specifically relevant to the same issues mentioned in #8649.

**Comment by jfkthame** [comment]
I think there's some confusion here regarding the term *hanging glyph*. In CSS Text 3, [8.2. Hanging Glyphs](https://drafts.csswg.org/css-text-3/#hanging) is discussing the convention of allowing certain characters to protrude or "hang" into the margin of the text block, so that their normal typographic advance is ignored (or placed outside the typographic bounds of the line). This is a deliberate typographic effect that involves adjusting the glyph's position from its "natural" place in the line.

The example of an italicized *f*
```
  <div class="figure" style="margin:0; font-size:10rem; font-family:garamond; font-style:italic">
    <span style="display:inline-block; width:1ch; height:1lh; background:green; opacity:0.5"></span><span style="vertical-align: top;">f</span><span style="display:inline-block; width:1ch; height:1lh; background:green; opacity:0.5"></span>
  </div>
```
mentioned in https://github.com/w3c/csswg-drafts/pull/9842#issuecomment-1992749312 is of a different nature. This glyph does not "hang" because it has been moved into the margin of the text block; it remains at its default inline position, occupying its normal space in the line. It has "hanging" ink because the actual glyph outline extends beyond its origin and advance (i.e. it has negative sidebearings). If it happens to occur at the beginning or end of a line, it will contribute ink overflow to the line because its negative sidebearings project into the margins, yet it is not a "hanging glyph" in the sense of CSS Text 3.

Such ink overflow can of course occur in the block as well as inline directions; an extreme example being "zalgo" text, where diacritics (that in themselves might not have any ink overflow) are stacked such that they render far beyond the typographic bounds of the line.

These glyph-level forms of ink overflow are conceptually quite distinct from the character-level ink overflow that CSS Text 3's hanging glyphs create, and the two cases should be addressed separately.

**Comment by chrishtr** [comment]
> These glyph-level forms of ink overflow are conceptually quite distinct from the character-level ink overflow that CSS Text 3's hanging glyphs create, and the two cases should be addressed separately.

I agree they are different. Are you just saying the two cases need separate spec text to account for them specially in defining the extent of ink overflow?

(Also, both contribute to ink overflow--do you agree?)

**Comment by jfkthame** [comment]
> > These glyph-level forms of ink overflow are conceptually quite distinct from the character-level ink overflow that CSS Text 3's hanging glyphs create, and the two cases should be addressed separately.
> 
> I agree they are different. Are you just saying the two cases need separate spec text to account for them specially in defining the extent of ink overflow?

Yes, to the extent that the spec aims to define/explain ink overflow, the two cases should be described separately; in particular, the current PR https://github.com/w3c/csswg-drafts/pull/9842 seems to use the italic-*f* example as though it were related to what CSS Text 3 calls "hanging glyphs", which I think is confusing.

> (Also, both contribute to ink overflow--do you agree?)

They do, though because of the different levels at which they operate, it's conceivable they might be handled differently by an implementation.

**Comment by szager-chromium** [comment]
@jfkthame As I'm sure is apparent, I am a novice when it comes to the details of text rendering. After reading the relevant specs, I understand the distinction you make between hanging glyphs and a glyph with a hanging outline. However, I can't seem to find an example or demonstration of a hanging glyph as described in the spec. `hanging-punctuation` has virtually no browser support, and `text-indent` does not generate ink overflow. Is there some other source of ink overflow from hanging glyphs that I am missing?

**Comment by frivoal** [comment]
Based on the fact that we resolved to close https://github.com/w3c/csswg-drafts/issues/8649, I think this and https://github.com/w3c/csswg-drafts/pull/9842 should be closed. @astearns , do you confirm?

**Comment by astearns** [comment]
In #8649 we resolved to add a note to overflow-3 about the platform differences in ink overflow. Was that ever done? Should there also be a note about these differences in text-3?

**Comment by fantasai** [comment]
I don't think it's necessary. The definition of what ink overflow is, and the fact that it include text rendering, is in css-overflow-3.

**Comment by frivoal** [comment]
Even though it's not marked as a note, @fantasai did include an editorial clarification along the lines of what we talked about for the note. I consider this addressed, and I don't think there's anything left to do. Closing. If a follow up is needed, I'd recommend opening specific individual issues.
